src/auth/identitypoolclient.ts

```diff
@@ -0,0 +1,218 @@
+   *   whether to retry on 401/403 API request errors.
+   */
+  constructor(
+    options: IdentityPoolClientOptions,
+    additionalOptions?: RefreshOptions
+  ) {
+    super(options, additionalOptions);
+    this.file = options.credential_source.file;
+    this.url = options.credential_source.url;
+    this.headers = options.credential_source.headers;
+    if (!this.file && !this.url) {
+      throw new Error('No valid Identity Pool "credential_source" provided');
+    }
+    // Text is the default format type.
+    this.formatType = options.credential_source.format?.type || 'text';
+    this.formatSubjectTokenFieldName =
+      options.credential_source.format?.subject_token_field_name;
+    if (this.formatType !== 'json' && this.formatType !== 'text') {
+      throw new Error(`Invalid credential_source format "${this.formatType}"`);
+    }
+    if (this.formatType === 'json' && !this.formatSubjectTokenFieldName) {
+      throw new Error(
+        'Missing subject_token_field_name for JSON credential_source format'
+      );
+    }
+  }
+
+  /**
+   * Triggered when a external subject token is needed to be exchanged for a GCP
+   * access token via GCP STS endpoint.
+   * This uses the `options.credential_source` object to figure out how
+   * to retrieve the token using the current environment. In this case,
+   * this either retrieves the local credential from a file location (k8s
+   * workload) or by sending a GET request to a local metadata server (Azure
+   * workloads).
+   * @return A promise that resolves with the external subject token.
+   */
+  async retrieveSubjectToken(): Promise<string> {
+    if (this.file) {
+      return await this.getTokenFromFile(
+        this.file!,
+        this.formatType,
+        this.formatSubjectTokenFieldName
+      );
+    } else {
+      return await this.getTokenFromUrl(
+        this.url!,
+        this.formatType,
+        this.formatSubjectTokenFieldName,
+        this.headers
+      );
+    }
+  }
+
+  /**
+   * Looks up the external subject token in the file path provided and
+   * resolves with that token.
+   * @param file The file path where the external credential is located.
+   * @param formatType The token file or URL response type (JSON or text).
+   * @param formatSubjectTokenFieldName For JSON response types, this is the
+   *   subject_token field name. For Azure, this is access_token. For text
+   *   response types, this is ignored.
+   * @return A promise that resolves with the external subject token.
+   */
+  private async getTokenFromFile(
+    filePath: string,
+    formatType: SubjectTokenFormatType,
+    formatSubjectTokenFieldName?: string
+  ): Promise<string> {
+    // Make sure there is a file at the path. lstatSync will throw if there is
+    // nothing there.
+    try {
+      // Resolve path to actual file in case of symlink. Expect a thrown error
+      // if not resolvable.
+      filePath = fs.realpathSync(filePath);
```

Comment:
using sync fs API in an async function does not seem right. On a network filesystem, these calls can block for considerable amount of time. Can we replace with async API e.g. `fs.realpath`, `fs.lstat`?

Reply:
Done